### PR TITLE
Added additional args to Exporter.exportChanges

### PR DESCRIPTION
**Pull request by simsta6**
- Added additional args to Exporter.exportChanges
- Added deprecation tag for old exportChanges overload
- Added deprecation tag for old IModelTransformer.transformChanges overload

### CODE REVIEW

File: packages/transformer/src/IModelExporter.ts [hunk 1]
```diff
@@ -81,7 +81,7 @@ export abstract class IModelExportHandler {
    *       This will become a part of onExportElement once that becomes async
    * @internal
    */
-  public async preExportElement(_element: Element): Promise<void> {}
+  public async preExportElement(_element: Element): Promise<void> { }
```
Comment (MichaelBelousov) [inline packages/transformer/src/IModelExporter.ts:84]: unnecessary change? Where does this come from? Was this a linter rule?

File: packages/transformer/src/IModelExporter.ts [hunk 2]
```diff
@@ -179,11 +179,18 @@ export class IModelExporter {
   private _sourceDbChanges?: ChangedInstanceIds;
   /**
    * Retrieve the cached entity change information.
-   * @note This will only be initialized after [IModelExporter.exportChanges] is invoked.
+   * @note This will be initialized after [IModelExporter.exportChanges] is invoked or it can be set.
```
Comment (MichaelBelousov) [inline packages/transformer/src/IModelExporter.ts:182]:
```suggestion
   * @note This may be set, or otherwise will be initialized after [IModelExporter.exportChanges] is invoked.
```

File: packages/transformer/src/IModelExporter.ts [hunk 3]
```diff
@@ -751,7 +760,7 @@ export class IModelExporter {
    * You may override this to load arbitrary json state in a transformer state dump, useful for some resumptions
    * @see [[IModelTransformer.loadStateFromFile]]
    */
-  protected loadAdditionalStateJson(_additionalState: any): void {}
+  protected loadAdditionalStateJson(_additionalState: any): void { }
```
Comment (MichaelBelousov) [inline packages/transformer/src/IModelExporter.ts:763]: is this the linter?

**Comment by MichaelBelousov** [comment]
what is the purpose of this change? What is a use case for setting it?

File: packages/transformer/src/IModelExporter.ts (hunk 3 above)
Comment (simsta6) [reply packages/transformer/src/IModelExporter.ts:763]: I formatted this file from vs code, so it may be the linter

**Comment by simsta6** [comment]
> what is the purpose of this change? What is a use case for setting it?

In our service, we need `ChangedInstanceIds` data. To reuse this data, I decided to add a public getter because adding a hook for this is not worth it.

**Comment by MichaelBelousov** [comment]
> In our service, we need `ChangedInstanceIds` data. To reuse this data, I decided to add a public getter because adding a hook for this is not worth it.

Maybe you can make it an option like `sourceChangeData`? I'd rather not expose a setter that can be used at any time.

File: packages/transformer/src/IModelExporter.ts [hunk 4]
```diff
@@ -268,9 +268,11 @@ export class IModelExporter {
    * @param user The user
    * @param startChangesetId Include changes from this changeset up through and including the current changeset.
    * If this parameter is not provided, then just the current changeset will be exported.
+   * @param changedInstanceIds Instance class that contains modified elements between 2 versions of an iModel.
+   * If this parameter is not provided, then [[ChangedInstanceIds.initialize]] will be called to discover changed elements.
    * @note To form a range of versions to export, set `startChangesetId` for the start (inclusive) of the desired range and open the source iModel as of the end (inclusive) of the desired range.
    */
-  public async exportChanges(user?: AccessToken, startChangesetId?: string): Promise<void> {
+  public async exportChanges(user?: AccessToken, startChangesetId?: string, changedInstanceIds?: ChangedInstanceIds): Promise<void> {
```
Comment (MichaelBelousov) [inline packages/transformer/src/IModelExporter.ts:275]: instead of adding a third plain argument, can it be an object containing this, so that future additional arguments can be added in a non-breaking way? Eventually I intend to deprecate the overload passing a string argument for `startChangesetId` and instead also put that in the object

Comment (MichaelBelousov) [reply packages/transformer/src/IModelExporter.ts:275]: It's not essential but yes, since it still introduces the new argument

Comment (simsta6) [reply packages/transformer/src/IModelExporter.ts:275]: We should start doing this as a standard practice

Comment (MichaelBelousov) [reply packages/transformer/src/IModelExporter.ts:275]: Yeah I'm willing to always bump minor on API changes, including non-breaking ones. It is better form.

Comment (simsta6) [reply packages/transformer/src/IModelExporter.ts]: Also converted processChanges to use object as argument

Comment (MichaelBelousov) [reply packages/transformer/src/IModelExporter.ts]: what you have now is a breaking change. I meant the following:

```ts
 public async exportChanges(user?: AccessToken, startChangesetId?: string, options: ExportChangesArgs): Promise<void> {
```

And then later we can deprecate that overload and do this. Although you can do so in this PR too, and we might as well if we're making it a minor release.

```ts
 /** @deprecated Don't use this overload */
 public async exportChanges(user?: AccessToken, startChangesetId?: string, options: ExportChangesArgs): Promise<void>
 /** desc */
 public async exportChanges(options: ExportChangesArgs): Promise<void>;
 public async exportChanges(userOrOptions?: AccessToken | ExportChangesArgs, startChangesetId?: string, options?: ExportChangesArgs): Promise<void> {
   const options = typeof userOrOptions === "string"
     ? {
       user: userOrOptions,
       startChangesetId,
       ...options
     }
     : userOrOptions; 
   //...
```

Comment (simsta6) [reply packages/transformer/src/IModelExporter.ts]: Great idea!

Comment (MichaelBelousov) [reply packages/transformer/src/IModelExporter.ts]: could you actually go ahead and add the deprecated overload as I showed above?

Comment (simsta6) [reply packages/transformer/src/IModelExporter.ts]: Added

Comment (MichaelBelousov) [reply packages/transformer/src/IModelExporter.ts]: do whatever works for VSCode, e.g. hovering later usages of `exportChanges` should show the description, but the invalid overload should be detected by both typescript (showing a ~strikethrough~ in vscode ) and eslint.

I believe it's really what I posted above. The full description remains on the valid overload signature which is declared first, then the deprecated overload is declared second with no description except `@deprecated in 0.1.x, use a single ExportChangesArgs object instead`, and that should work. Unless I removed them, I believe I've done this in other places.

Comment (simsta6) [reply packages/transformer/src/IModelExporter.ts]: I changes descriptions. I was using itwinjs core lib as example.

Comment (MichaelBelousov) [reply packages/transformer/src/IModelExporter.ts]: do you think you could cherry-pick in my version where I did something similar? 2f1189eafa68c7382c7e220030961157f487599a
You might also need c2f2da9d400f34e7b8b1c6b2fcd5d07fdae537a4...

Comment (simsta6) [reply packages/transformer/src/IModelExporter.ts]: Is these changes still not in master? Maybe it would be better just merge this PR or close it and start over again

Comment (simsta6) [reply packages/transformer/src/IModelExporter.ts]: There is multiple issues with those commits:
 - using types like `ChangesetIndexOrId` makes hardly readable code
 - importing types from different file, that can be reused, but actually is for different method is also makes code look like spaghetti

I would rather stick with my current version

Comment (MichaelBelousov) [reply packages/transformer/src/IModelExporter.ts]: well the types are highly related, those functions call each other so they should have a relatively consistent API. And I disagree that `ChangesetIndexOrId` makes for hardly readable code. The value of using `ChangesetIndexOrId` is that the user can pass what they have without having to resolve the other kind using a query. This change actually reduced the amount of querying the transformer does which needs an index but was requesting an id.

If you're referring to the new "args" declarations as spaghetti, they are slightly unfortunate but that's a natural reality of supporting multiple overloads and makes sense until the next major version when we can drop the deprecated overload.

Are you still against this?

Comment (simsta6) [reply packages/transformer/src/IModelExporter.ts]: I would rather have multiple types or interfaces which extends basic ones. Because there is a possibility to have a need to add additional property that will be used in one function and not in other function

Comment (MichaelBelousov) [reply packages/transformer/src/IModelExporter.ts]: maybe I forgot a commit, there should be a second type that extends from the base.

Comment (MichaelBelousov) [reply packages/transformer/src/IModelExporter.ts]: I think I forgot [ff19036f](https://github.com/iTwin/imodel-transformer/commit/ff19036ec6456c59891f6d56bdb2982e082d8d2f)
It's really a subset of federation-guid-optimization branch, which you could extract using
`git add -p federation-guid-optimization packages/transformer/src/IModelTransformer.ts packages/transformer/src/IModelExport.ts` and just ignore all other changes. Or try to cherry-pick those 3 commits but I didn't thoroughly check if that contains all the changes.

Comment (MichaelBelousov) [reply packages/transformer/src/IModelExporter.ts]: I created #55 to demonstrate what I'd prefer

File: packages/transformer/src/IModelExporter.ts [hunk 5]
```diff
@@ -264,29 +278,48 @@ export class IModelExporter {
...
-   */
-  public async exportChanges(user?: AccessToken, startChangesetId?: string): Promise<void> {
-    if (!this.sourceDb.isBriefcaseDb()) {
+  /**
+   * Export changes from the source iModel.
+   */
+  public async exportChanges(args?: ExportChangesOptions): Promise<void>;
+  /** @deprecated in 0.1.x, use a single ExportChangesArgs object instead */
+  public async exportChanges(accessToken?: AccessToken, startChangesetId?: string, args?: ExportChangesOptions): Promise<void>;
+  /** @internal Don't use this overload. */
```
Comment (MichaelBelousov) [inline packages/transformer/src/IModelExporter.ts:287]: nit: is it necessary to mark this as internal? It's not an overload but the implementation. I think typescript will identify similar usage (more than 1 argument) as the deprecated overload so this is unnecessary

**Review by MichaelBelousov — APPROVED** [review]
need to fix lint errors but LGTM!

File: packages/transformer/src/IModelExporter.ts (hunk 5 above)
Comment (simsta6) [reply packages/transformer/src/IModelExporter.ts]: Okay, I removed it :)
